drivers/at24cxxx/at24cxxx.c

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_at24cxxx
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the AT24C128 / AT24C256 EEPROM units.
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ * @}
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+
+#define ENABLE_DEBUG  (0)
+#include "debug.h"
+#include "assert.h"
+#include "xtimer.h"
+
+#include "at24cxxx_defines.h"
+#include "at24cxxx.h"
+
+#ifndef MIN
+#define MIN(a, b) ((a) > (b) ? (b) : (a))
+#endif
+
+/**
+ * @brief Calculate x mod y, if y is a power of 2
+ */
+#define MOD_POW2(x, y) ((x) & ((y) - 1))
+
+/**
+ * @brief I2C bus number shortcut
+ */
+#define I2C_BUS                     (dev->params.i2c)
+/**
+ * @brief Pin wp shortcut
+ */
+#define PIN_WP                      (dev->params.pin_wp)
+/**
+ * @brief EEPROM size shortcut
+ */
+#define EEPROM_SIZE                 (dev->params.eeprom_size)
+/**
+ * @brief I2C device address shortcut
+ */
+#define I2C_ADDR                    (dev->params.dev_addr)
+/**
+ * @brief Page size shortcut
+ */
+#define PAGE_SIZE                   (dev->params.page_size)
+/**
+ * @brief Word address length shortcut
+ */
+#define WORD_ADDR_LEN               (dev->params.word_addr_len)
+/**
+ * @brief Max polls shortcut
+ */
+#define MAX_POLLS                   (dev->params.max_polls)
+
+#ifndef AT24CXXX_SET_BUF_SIZE
+/**
+ * @brief  Adjust to configure buffer size
+ */
+#define AT24CXXX_SET_BUF_SIZE       (PAGE_SIZE)
+#endif
+
+static
+int _read(const at24cxxx_t *dev, uint32_t pos, void *data, size_t len)
+{
+    int check;
+    uint8_t polls = MAX_POLLS;
+    uint8_t dev_addr;
+    uint8_t flags = 0;
+
+    if (WORD_ADDR_LEN > 11) {
+        /* 2 bytes word address length*/
+        /* append page address bits to device address (if any) */
+        dev_addr  = (I2C_ADDR | ((pos & 0xFF0000) >> 16));
+        pos &= 0xFFFF;
+        flags = I2C_REG16;
+    }
+    else {
+        /* append page address bits to device address (if any) */
+        dev_addr = (I2C_ADDR | ((pos & 0xFF00) >> 8));
+        pos &= 0xFF;
+    }
```

Comment:
This needs more explaining. Reading the code, it seems that the I2C device address and the memory address are multiplexed. But that seems like a very strange idea.

Reply:
OK, the AT24C1024 seems to be something like two AT24C512 under the hood, where one address bit selects which AT24C512 is talked to.

Maybe we can just drop support for that one to get the code a bit easier to read?

(I think the driver would still allow using the full 128KiB of an AT24C1024 by treating it as two AT24C512 with separate addresses...)

Reply:
Any `at24cxxx` has a 7 bit i2c address, `1010???`, where `?` can be address specific values, or the most significant bits of the address of the word you address in the EEPROM memory. The devices also differ in the number of bytes to address a data word in memory, either 1 or 2 bytes.

`Example with at24c16a:`
It has 2048 byte of memory and `11 bits address length`, but it used only `1 byte as word address`.
To achieve 11 bit, it puts the three most significant bits into the device address (`p2, p1, p0`).
The address looks like this: `1 0 1 0 p2 p1 p0`. This comes at a cost that you don´t have address bits (`A3, A2, A0`), so you can only have one at24c16a attached to an i2c bus.

code:
We have a 32 bit integer which could use all 11 bits of addressing.
So the device address `1 0 1 0 0 0 0 0 `becomes `OR`´ed with the bits 15 - 8 from the actual position in memory. So if the bist 10, 9 or 8 are 1 they become appended to the device address as p2, p1 and p0. `And i2c_write_regs()` takes the bits 7 to 0 as register address.

If an at24cxxx had more than 11 word address bits, it would destroy the address prefix 1010, so any type with more than 11 address bits uses 2 bytes for the word address.

Example for the` at24c1024`:
131072 bytes memory,` 17 bit word address width`.
It´s address width is too large for 2 bytes, so it puts the most significant bit in the device address.

Code:
We get a 32 bit integer where 17 bit are the word address.
So the device address `1 0 1 0 0 A1 0 `becomes OR´ed with bits 23 to 16 from the position.
It is true that for the bits 23 to 16, only bit 16 could be one, because it is checked that `pos + len` fits in the EEPROM´s size.
And i2c_write_regs() takes bits 15-0 as the register address